crates/runtime/src/instance/allocator/pooling/uffd.rs

```diff
@@ -0,0 +1,833 @@
+
+            return Some(AddressLocation::StackPage {
+                page_addr: (stack_start + page_offset) as _,
+                len: self.page_size,
+                index,
+                guard_page: stack_offset < self.page_size,
+            });
+        }
+
+        None
+    }
+}
+
+unsafe fn wake_guard_page_access(
+    uffd: &Uffd,
+    page_addr: *const u8,
+    len: usize,
+) -> Result<(), String> {
+    // Set the page to NONE to induce a SIGSEV for the access on the next retry
+    region::protect(page_addr, len, region::Protection::NONE)
+        .map_err(|e| format!("failed to change guard page protection: {}", e))?;
+
+    uffd.wake(page_addr as _, len).map_err(|e| {
+        format!(
+            "failed to wake page at {:p} with length {}: {}",
+            page_addr, len, e
+        )
+    })?;
+
+    Ok(())
+}
+
+unsafe fn initialize_wasm_page(
+    uffd: &Uffd,
+    instance: &Instance,
+    page_addr: *const u8,
+    memory_index: usize,
+    page_index: usize,
+) -> Result<(), String> {
+    if let Some(MemoryInitialization::Paged { page_size, map }) =
+        &instance.module.memory_initialization
+    {
+        let memory_index = DefinedMemoryIndex::new(memory_index);
+        let memory = instance.memory(memory_index);
+        let pages = &map[memory_index];
+        debug_assert_eq!(WASM_PAGE_SIZE % page_size, 0);
+
+        let count = WASM_PAGE_SIZE / page_size;
+        let start = page_index * count;
+
+        for i in start..start + count {
+            let dst = memory.base.add(i * page_size);
+
+            match pages.get(i) {
+                Some(Some(data)) => {
+                    log::trace!(
+                        "copying page initialization data from {:p} to {:p} with length {}",
+                        data,
+                        dst,
+                        page_size
+                    );
+
+                    // Copy the page data without waking
+                    uffd.copy(data.as_ptr() as _, dst as _, *page_size, false)
+                        .map_err(|e| {
+                            format!(
+                                "failed to copy page from {:p} to {:p} with length {}: {}",
+                                data, dst, page_size, e
+                            )
+                        })?;
+                }
+                _ => {
+                    log::trace!("zeroing page at {:p} with length {}", dst, page_size);
+
+                    // No data, zero the page without waking
+                    uffd.zeropage(dst as _, *page_size, false).map_err(|e| {
+                        format!(
+                            "failed to zero page at {:p} with length {}: {}",
+                            dst, page_size, e
+                        )
+                    })?;
+                }
+            }
+        }
+
+        // Finally wake the entire wasm page
+        uffd.wake(page_addr as _, WASM_PAGE_SIZE).map_err(|e| {
+            format!(
```

Comment:
FWIW I find the error-handling boilerplate here to take up a lot of screen real-estate, it'd be nice to use perhaps a helper function to create these messages to make the `map_err` format onto a single line if possible.

Reply:
I can do that or instead make use of `anyhow`.